### ISSUE TITLE
My data doesnt prevails on my sql azure database, after login

### ISSUE THREAD

**Issue by Quico20** [Azure/azure-mobile-apps-net-client#504]
I'm using Azure Mobile App SDK, and I have 4 tables Account, personal, Labor and Subscription.
I'll leave the code here of my models:

`public class Account : TableData
    {
        [JsonProperty(PropertyName ="userid")]
        public string UserId { get; set; }
        [JsonProperty(PropertyName = "personalid")]
        public string PersonalId { get; set; }
        [JsonProperty(PropertyName = "subscriptionid")]
        public string SubscriptionId { get; set; }
        [JsonProperty(PropertyName = "subscriptionappleid")]
        public string SubscriptionAppleId { get; set; }
        [JsonProperty(PropertyName = "subscriptiongoogleid")]
        public string SubscriptionGoogleId { get; set; }
        [JsonProperty(PropertyName = "laborid")]
        public string LaborId { get; set; }

        public override bool Equals(object obj)
        {
            return Equals<Account>(obj, this);
        }

        public override int GetHashCode()
        {
            return Id.GetHashCode();
        }
    }`

`public class Personal : TableData
    {
        [Required]
        [StringLength(50)]
        [JsonProperty(PropertyName = "name")]
        public string Name { get; set; }
        [Required]
        [StringLength(50)]
        [JsonProperty(PropertyName = "dadlastname")]
        public string DadLastName { get; set; }
        [Required]
        [StringLength(50)]
        [JsonProperty(PropertyName = "momlastname")]
        public string MomLastName { get; set; }
        [Required]
        [StringLength(50)]
        [JsonProperty(PropertyName = "email")]
        public string Email { get; set; }
        [Required]
        [JsonProperty(PropertyName = "sex")]
        public Gender Sex { get; set; }
        [Required]
        [Column(TypeName = "datetime2")]
        [JsonProperty(PropertyName = "birthdate")]
        public DateTime? BirthDate { get; set; }

        public override bool Equals(object obj)
        {
            return Equals<Personal>(obj, this);
        }

        public override int GetHashCode()
        {
            return Id.GetHashCode();
        }
    }`

`public class Subscription : TableData
    {
        [Required]
        [JsonProperty(PropertyName = "subscriptiontype")]
        public SubscriptionType SubscriptionType { get; set; }
        [Required]
        [Column(TypeName = "datetime2")]
        [JsonProperty(PropertyName = "subscriptiondate")]
        public DateTime? SubscriptionDate { get; set; }
        [Required]
        [Column(TypeName = "datetime2")]
        [JsonProperty(PropertyName = "expirationdate")]
        public DateTime? ExpirationDate { get; set; }
        [Required]
        [JsonProperty(PropertyName = "standardsubscriptiontype")]
        public SubscriptionType StandardSubscriptionType { get; set; }
        [Required]
        [Column(TypeName = "datetime2")]
        [JsonProperty(PropertyName = "standardsubscriptiondate")]
        public DateTime? StandardSubscriptionDate { get; set; }
        [Required]
        [Column(TypeName = "datetime2")]
        [JsonProperty(PropertyName = "standardexpirationdate")]
        public DateTime? StandardExpirationDate { get; set; }

        public override bool Equals(object obj)
        {
            return Equals<Subscription>(obj, this);
        }

        public override int GetHashCode()
        {
            return Id.GetHashCode();
        }
    }`

now as you can see Account model has Foreing Keys to Personal, Labor and Subscription.
whenever I make a change on Personal, labor or susbcription i updated my table and I can see at my cloud that the information has been changed, for example.

at Personal model I make a change on my birthdate from Oct 5th, 1981 to Oct. 6th 1981. and it updates successfully, but if I close my app and after 30 mins or 20 minutes I open the app again as soon as I login using the login method who has nothing to do with Sql Azure, the database information goes back to the previous values, and I have to wait like 20 or 30 minutes for this to happend I close my app and the reopen it inmediately the database or the record doesnt get affected.

Now this only happens with the table I mentioned before, If I make edition to other tables, they dont get affected by the situation I mentioned before.

Am I doing something wrong??? this didnt happened before, I have tried retiring offline support for this tables thinking that might be an issue with sqlLite but the issue still prevails.

I ignore the reason, of why is this happening.

Hopefully you can help me with some ideas, I dont have any clue about it.
Should I get rid of Foreing Keys???Should I delete the whole record and create a new one a reassign a new PersonalId or LaborId to the Account Model??? as a workaround.

**Comment by Jamminroot** [comment]
Hi, @Quico20, thanks for getting in touch.
Correct me if im wrong:
When you **update** data in your `Personal` table, data in `Personal` table gets reverted back on relogin, but that only occurs only with this table. Is that correct?
Also, you mentioned foreign keys, but I can't see any connections in the code you provided. Can you elaborate on this point?
A short snippet on how exactly you modify data would be helpful to resolve your problem.

**Comment by Quico20** [comment]
Ok, the foreing key are declared at my backend project. I'm gonna show you my backend models, according to Adrian Halls Azure Book he recommended to create the models at the backend project like the code i'm just putting here:
                                                                                                               
       public class Account : EntityData      
       {
        public string UserId { get; set; }

        #region Relationship with Personal
        public string PersonalId { get; set; }

        [ForeignKey(nameof(PersonalId))]
        public Personal PersonalInfo { get; set; }
        #endregion

        #region Relationship with Subscription
        public string SubscriptionId { get; set; }

        [ForeignKey(nameof(SubscriptionId))]
        public Subscription SubscriptionInfo { get; set; }
        #endregion

        #region Relationship with Labor
        public string LaborId { get; set; }

        [ForeignKey(nameof(LaborId))]
        public Labor LaborInfo { get; set; }
        #endregion

        #region Relationship with Subscription Apple
        public string SubscriptionAppleId { get; set; }

        [ForeignKey(nameof(SubscriptionAppleId))]
        public SubscriptionApple SubscriptionAppleInfo { get; set; }
        #endregion

        #region Relationship with Subscription Google
        public string SubscriptionGoogleId { get; set; }

        [ForeignKey(nameof(SubscriptionGoogleId))]
        public SubscriptionGoogle SubscriptionGoogleInfo { get; set; }
        #endregion
      }

as you can see the server model of account has the ForeignKey Attribute but at the Client model since I use SqLite i have to flat the model, because as I recall SqLite doesnt support 1:way relationships diffently according to [Adrian Hall book](https://adrianhall.github.io/develop-mobile-apps-with-csharp-and-azure/chapter3/relationships/).

now I'm gonna show you the process involved at resetting the table and I'll add screenshots if that serves the purpose of this isssue, this method is trigget whenever i click on the facebook login button. and my database resets after the line with resets text has been executed.

             private async Task ExecuteLoginCommand()
             {
              SetBusy(IsBusy, GetLocalizeString("PagesViewModelBusySyncWaitingMessage"));
               try
               {               
                //For multiple providers authentication.
                //ProviderService.Provider = MobileServiceAuthenticationProvider.Facebook;
                await App.CloudService.LoginAsync(); //RESETS AFTER EXECUTE
                Table = await App.CloudService.GetOnlineTableAsync<Account>().ReadAllItemsAsync();
                if (Table.Count > 0)
                {
                    await NavigationService.NavigateAsync(typeof(MainMenuPage));
                }
                else
                {
                    Debug.WriteLine($"[Login] - No existe la cuenta registrada en Ceneam");
                    await NavigationService.NavigateAsync(typeof(RegisterPage));
                }                
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"[Login] Error = {ex.Message}");
            }
            finally
            {
                SetBusy(!IsBusy);
            }
        }

My LoginAsync Method looks like this:

           public async Task<MobileServiceUser> LoginAsync()
           {
                
            ILoginUWPProvider UwpProvider = new LoginUWPProvider();

            client.CurrentUser = UwpProvider.RetrieveTokenFromSecureStore();
            if (client.CurrentUser != null)
            {
                // User has previously been authenticated - try to Refresh the token
                try
                {
                    var refreshedUser = await client.RefreshUserAsync();
                    if (refreshedUser != null)
                    {
                        UwpProvider.StoreTokenInSecureStore(refreshedUser);
                        return refreshedUser;
                    }
                }
                catch (Exception refreshedException)
                {
                    Debug.WriteLine($"Could not refreshed token: {refreshedException.Message}");
                }
            }

            if (client.CurrentUser != null && !IsTokenExpired(client.CurrentUser.MobileServiceAuthenticationToken))
            {
                //User has previously been authenticated, no refresh is required
                return client.CurrentUser;
            }

            // We need to ask for credentials at this point
            await UwpProvider.LoginAsync(client);

            if(client.CurrentUser !=null)
            {
                // We were able to successfully log in
                UwpProvider.StoreTokenInSecureStore(client.CurrentUser);
            }

            return client.CurrentUser;
        }

Ok now I'll make a Debug and add screenshots of my database at Labor table before I login and then after I login, stepping method by method.

![image](https://user-images.githubusercontent.com/13721058/58431270-a0911300-8072-11e9-90c2-e013b23350d6.png)
 as you can see at the picture check the SalaryLevel column at the last record it says 80, after I login or after the LoginAsync method is executed it returns to 76. that was the last value it has as of yesterday night.

as soon as I Refresh the token at my facebook login it throws me this execption message:

> Exception thrown: 'Microsoft.WindowsAzure.MobileServices.MobileServiceInvalidOperationException' in System.Private.CoreLib.ni.dll
Could not refreshed token: Refresh failed with a 400 Bad Request error. The identity provider does not support refresh, or the user is not logged in with sufficient permission.

and it resets the table again, I just found the exact point where it take the table back to the previous value. this is the screenshot of my table as of right now 11:38am local time Mexico Center time.

![image](https://user-images.githubusercontent.com/13721058/58431551-fe722a80-8073-11e9-8ba8-6b4341dd68d8.png)

see how the data changed back. now this never happened and it only affect the Tables who has a relationship such as Personal, Labor or Subscription with are linked to Account Model through Foreign Keys. Next time I'll try by removing the refreshing method since facebook and twitter doesnt support it. but i'll have to wait for 30 minutes, since this behavior show After I login like 20 to 30 minutes after using the UWP App.

Hopefully It help to find out more about this issue.

**Comment by Quico20** [comment]
I had just made the test again this time by deleting this code:

```
        if (client.CurrentUser != null)
        {
            // User has previously been authenticated - try to Refresh the token
            try
            {
                var refreshedUser = await client.RefreshUserAsync();
                if (refreshedUser != null)
                {
                    UwpProvider.StoreTokenInSecureStore(refreshedUser);
                    return refreshedUser;
                }
            }
            catch (Exception refreshedException)
            {
                Debug.WriteLine($"Could not refreshed token: {refreshedException.Message}");
            }
        }
```
since the last time it reseted back to previous values, but it didnt work as soon as I called this method at my Login Event:

```
        private async Task ExecuteLoginCommand()
        {
            SetBusy(IsBusy, GetLocalizeString("PagesViewModelBusySyncWaitingMessage"));
            try
            {
                //For multiple providers authentication.
                //ProviderService.Provider = MobileServiceAuthenticationProvider.Facebook;
                await App.CloudService.LoginAsync();
                //RESET VALUES AS SOON AS THE ACCOUNT TABLE IS CALL
                Table = await App.CloudService.GetOnlineTableAsync<Account>().ReadAllItemsAsync();
                if (Table.Count > 0)
                {
                    await NavigationService.NavigateAsync(typeof(MainMenuPage));
                }
                else
                {
                    Debug.WriteLine($"[Login] - No existe la cuenta registrada en Ceneam");
                    await NavigationService.NavigateAsync(typeof(RegisterPage));
                }
                
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"[Login] Error = {ex.Message}");
            }
            finally
            {
                SetBusy(!IsBusy);
            }
        }
```
I think next time I'll try to get rid of Foreign Keys now, and make the links on my own manually. since they are the only tables with 1:way relationship it should solve it for now, because I dont see any trouble with my other tables that has no relationships at all.

This is a awful bug it never happened before, it just began to happen recently, because I use to changed the Labor base salary through another table that update base salary values, and it kept changes, but now even that method has been affected, I think it has something to do with the SDK, Azure Mobile Apps Net Client SDK.

if you need more information ask me I'll try to give you as much information as I can gather.

Hopefully you can find an answer to this issue soon.

**Comment by Quico20** [comment]
I had just try by deleting the Foreign keys at my server side model, the result still the same,  I think the only difference between Personal, Labor and Subscription tables are, that they doesnt have a UserId column as the other tables. Tomorrow I'll try by adding uniqueness to it. if it works I'll have to modify my swapstation table as well.

**Comment by Jamminroot** [comment]
Ok. we'll wait for your update. We are not closing the thread yet

**Comment by Quico20** [comment]
Firstable I just solved it. I want to apologize because it was MY MISTAKE, it turns out that I hard coded my account data at the Configuriation seeding method of  my cloud which run everytime someone login, the Azure Mobile Apps SDK has no flaws on it.

I had just deleted the data that rollback my information, or better said to updated my information.
It was a hard error to find because my code was fine, it was the behavior of my cloud that was wrong due to hardcoded the account information because I had an issue months ago, were i need to hardcode my account information in my cloud.

anyways I think this issue can be close now!!!
